doc/libraries.rst

```diff
@@ -71,12 +71,26 @@ An ESP8266 port of SoftwareSerial library done by Peter Lerup (@plerup) supports
 ESP-specific APIs
 -----------------
 
-Some ESP-specific APIs related to deep sleep, RTC and flash memories are available in the ``ESP`` object.
+Some ESP-specific APIs related to the deep, modem, and light sleep modes, RTC and flash memory are available in the ``ESP`` object.
 
 ``ESP.deepSleep(microseconds, mode)`` will put the chip into deep sleep. ``mode`` is one of ``WAKE_RF_DEFAULT``, ``WAKE_RFCAL``, ``WAKE_NO_RFCAL``, ``WAKE_RF_DISABLED``. (GPIO16 needs to be tied to RST to wake from deepSleep.) The chip can sleep for at most ``ESP.deepSleepMax()`` microseconds. If you implement deep sleep with ``WAKE_RF_DISABLED`` and require WiFi functionality on wake up, you will need to implement an additional ``WAKE_RF_DEFAULT`` before WiFi functionality is available.
 
 ``ESP.deepSleepInstant(microseconds, mode)`` works similarly to ``ESP.deepSleep`` but  sleeps instantly without waiting for WiFi to shutdown.
 
+``ESP.forcedModemSleep(microseconds, callback)`` immediately puts the chip into forced ``MODEM_SLEEP``. A microseconds duration after which the sleep mode returns to the automatic sleep mode that was effective before this call can be given, a value of 0 or 0xFFFFFFF turns off that timeout. The optional callback function will be invoked when the forced modem sleep ends.
+
+``ESP.forcedModemSleepOff()`` immediately returns the chip to the automatic sleep mode in effect before the preceeding call to ``ESP.forcedModemSleep``.
+
+``ESP.forcedLightSleepBegin(microseconds, callback)`` works in tandem with ``ESP.forcedLightSleepEnd(cancel)`` to put the chip into forced ``LIGHT_SLEEP``. A microseconds duration after which the sleep mode returns to the automatic sleep mode that was effective before this call can be given, a value of 0 or 0xFFFFFFF turns off that timeout. The optional callback function will be invoked when the forced light sleep ends. Forced light sleep halts the CPU, in addition to the timeout, it can be awakened via GPIO input. Between the calls to ``ESP.forcedLightSleepBegin`` and ``ESP.forcedLightSleepEnd`` any GPIOs except GPIO16 to use for wakeup can be set up.
+
+``ESP.forcedLightSleepEnd(cancel)`` causes the chip to enter the forced light sleep mode that was set up as described for ``ESP.forcedLightSleepBegin``. The optional cancel argument, if true, prevents the sleep mode transition from occuring and returns to the automatic sleep mode that was effective before ``ESP.forcedLightSleepBegin``.
```

Comment:
Done, I think, thanks for pointing out what was not easy to understand.
> Light, modem fpm sleep somehow change wifi driver sleep setting?

Please be more specific, what is the nature of your question?

Reply:
>Since you have already tested this... does this mean:
    light sleep is entered at the end of SYS idle task
    light sleep is entered immediately with wifi_fpm_do_sleep()

Forced Light Sleep is entered sometime _during_ the idle task, and it's a fuzzy number between 2 and 8ms.  Most chips entered Sleep right around 2ms, but I had one chip that needed 8 myself last year.  In my version I bumped that idle to 10ms in case there was a chip that needed even longer than the one weird chip I had.  I didn't want to increase it too much as the whole point of Forced Light Sleep is that it wakes much faster than Deep Sleep, albeit at a much higher power consumption during Sleep (460uA vs 20uA for Deep Sleep).